sound/soc/sof/ipc4.c

```diff
@@ -0,0 +1,218 @@
+	{150, "Attempt to configure SCLK while I2S port is running"},
+	{151, "Attempt to configure MCLK while I2S port is running"},
+	{152, "Attempt to stop SCLK that is not running"},
+	{153, "Attempt to stop MCLK that is not running"},
+	{160, "Reserved (ADSP_IPC_PIPELINE_NOT_INITIALIZED removed)"},
+	{161, "Reserved (ADSP_IPC_PIPELINE_NOT_EXIST removed)"},
+	{162, "Reserved (ADSP_IPC_PIPELINE_SAVE_FAILED removed)"},
+	{163, "Reserved (ADSP_IPC_PIPELINE_RESTORE_FAILED removed)"},
+	{165, "Reserved (ADSP_IPC_PIPELINE_ALREADY_EXISTS removed)"},
+};
+
+void sof_ipc4_check_reply_status(struct snd_sof_dev *sdev, u32 msg)
+{
+	u32 status = msg & SOF_IPC4_REPLY_STATUS_MASK;
+	int i;
+
+	sdev->ipc->msg.reply_error = status;
+	if (status) {
+		for (i = 0; i < ARRAY_SIZE(ipc4_status); i++) {
+			if (ipc4_status[i].status == status) {
+				dev_err(sdev->dev, "FW reported error: %s", ipc4_status[i].msg);
+				return;
+			}
+		}
+
+		dev_err(sdev->dev, "FW reported unknown error, status = %d", status);
+	}
+
+	snd_sof_ipc_reply(sdev, msg);
+}
+EXPORT_SYMBOL(sof_ipc4_check_reply_status);
+
+/* wait for IPC message reply */
+static int ipc4_tx_wait_done(struct snd_sof_ipc *ipc, struct snd_sof_ipc_msg *msg,
+			     void *reply_data)
+{
+	struct snd_sof_dev *sdev = ipc->sdev;
+	int ret;
+
+	/* wait for DSP IPC completion */
+	ret = wait_event_timeout(msg->waitq, msg->ipc_complete,
+				 msecs_to_jiffies(sdev->ipc_timeout));
+	if (ret == 0) {
+		dev_err(sdev->dev, "error: ipc timed out for header:0x%x extension:0x%x",
+			msg->header, msg->extension);
+		return -ETIMEDOUT;
+	} else if (msg->reply_error) {
+		dev_err(sdev->dev, "error: ipc error for msg 0x%x : 0x%x\n",
+			msg->header, msg->extension);
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int sof_ipc4_tx_message_unlocked(struct snd_sof_ipc *ipc, u32 header,
+					u32 extension, void *msg_data, size_t msg_bytes,
+					void *reply_data, size_t reply_bytes)
+{
+	struct snd_sof_dev *sdev = ipc->sdev;
+	struct snd_sof_ipc_msg *msg;
+	int ret;
+
+	if (ipc->disable_ipc_tx)
+		return -ENODEV;
+
+	/*
+	 * The spin-lock is also still needed to protect message objects against
+	 * other atomic contexts.
+	 */
+	spin_lock_irq(&sdev->ipc_lock);
+
+	/* initialise the message */
+	msg = &ipc->msg;
+
+	msg->header = header;
+	msg->extension = extension;
+	msg->msg_size = msg_bytes;
+	msg->reply_size = reply_bytes;
+	msg->reply_error = 0;
+
+	/* attach any data */
+	if (msg_bytes > msg->msg_data_size)
+		return -ENOMEM;
+
+	memcpy(msg->msg_data, msg_data, msg_bytes);
+
+	sdev->msg = msg;
+
+	ret = snd_sof_dsp_send_msg(sdev, msg);
+	/* Next reply that we receive will be related to this message */
+	if (!ret)
+		msg->ipc_complete = false;
+
+	spin_unlock_irq(&sdev->ipc_lock);
+
+	if (ret < 0) {
+		dev_err_ratelimited(sdev->dev, "error: ipc tx failed with error %d", ret);
+		return ret;
+	}
+
+	/* now wait for completion */
+	if (!ret)
+		ret = ipc4_tx_wait_done(ipc, msg, reply_data);
+
+	return ret;
```

Comment:
simply:
return ipc4_tx_wait_done();

Btw: do we not want to have similar logging as with IPC3?

Reply:
We may do it later.